core/client/fs/src/main/java/alluxio/client/metrics/MetricsHeartbeatContext.java

```diff
@@ -0,0 +1,241 @@
+import alluxio.master.MasterInquireClient;
+import alluxio.util.ThreadFactoryUtils;
+import alluxio.util.network.NetworkAddressUtils;
+
+import com.google.common.base.Preconditions;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A class used to track metrics heartbeats to a master.
+ *
+ * The class should be instantiated when a new FileSystemContext is created with a
+ * configuration that points to a given master. As new FileSystemContexts are created, if they
+ * connect to the same master RPC address, then they can simply be added to this context so
+ * that their information is included in the metrics heartbeat.
+ *
+ * For each separate master RPC address, a new instance of this class is created. As
+ * FileSystemContexts are closed, they remove their application Ids from the internal metrics
+ * heartbeat. When a context reaches 0 applicationIds it will automatically close and remove
+ * itself from the internal MASTER_METRICS_HEARTBEAT map.
+ *
+ * When the final FileSystemContext closes and removes its heartbeat from metrics it will also
+ * shutdown and close the executor service until a new FileSystemContext is created.
+ */
+@ThreadSafe
+public class MetricsHeartbeatContext {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsHeartbeatContext.class);
+
+  /** A map from master RPC address to heartbeat context instances. */
+  private static final ConcurrentHashMap<String, MetricsHeartbeatContext>
+      MASTER_METRICS_HEARTBEAT = new ConcurrentHashMap<>();
+
+  /** A value that tracks whether or not we've registered the shutdown hook. */
+  private static boolean sAddedShudownHook = false;
+
+  /** The service which executes metrics heartbeats. */
+  private static ScheduledExecutorService sExecutorService;
+
+  private final String mMasterAddr;
+  private final MetricsMasterClient mMetricsMasterClient;
+  private final ClientMasterSync mClientMasterSync;
+  private ScheduledFuture<?> mMetricsMasterHeartbeatTask;
+  private final AtomicInteger mCtxCount;
+  private final AlluxioConfiguration mConf;
+
+  private MetricsHeartbeatContext(String masterAddr, ClientContext ctx,
+      MasterInquireClient inquireClient) {
+    mMasterAddr = masterAddr;
+    mCtxCount = new AtomicInteger(0);
+    mConf = ctx.getConf();
+    // setup metrics master client sync
+    mMetricsMasterClient = new MetricsMasterClient(MasterClientContext
+        .newBuilder(ctx)
+        .setMasterInquireClient(inquireClient)
+        .build());
+    mClientMasterSync = new ClientMasterSync(mMetricsMasterClient, mConf);
+  }
+
+  private synchronized void addContext(String appId) {
```

Comment:
Why is this called `addContext`?

Reply:
It was called addContext because it's called when a FileSystemContext adds a heartbeat

Reply:
It is a bit confusing because we take an `appId` as the argument, how about something like `registerApplication` or `addApplication`?

Reply:
Technically this is outdated now as it doesn't even take an appId anymore. @aaudiber and I decided it would be best to only allow 1 appId per-JVM because MetricsSystem is still `static`. So now it is no-arg.

As of now, its only function is to track `FileSystemContext`s through `addContext` or `removeContext`. I would like to keep it the way it is for now, and we may change it in the future.